File: app/src/main/java/com/u1fukui/bbs/view/activity/NotificationListActivity.java
package com.u1fukui.bbs.view.activity;


import android.databinding.DataBindingUtil;
import android.databinding.ObservableList;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.view.ViewGroup;

import com.u1fukui.bbs.R;
import com.u1fukui.bbs.databinding.ActivityNotificationListBinding;
import com.u1fukui.bbs.databinding.ViewNotificationCellBinding;
import com.u1fukui.bbs.view.customview.BindingHolder;
import com.u1fukui.bbs.view.customview.ObservableListRecyclerAdapter;
import com.u1fukui.bbs.viewmodel.NotificationListViewModel;
import com.u1fukui.bbs.viewmodel.NotificationViewModel;

import javax.inject.Inject;

public class NotificationListActivity extends BaseActivity {

    @Inject
    NotificationListViewModel viewModel;

    private ActivityNotificationListBinding binding;

    private Adapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_notification_list);
        binding.setViewModel(viewModel);

        initToolbar(binding.toolbar, true);
        initViews();

        viewModel.start();
    }

    private void initViews() {
        adapter = new Adapter(viewModel.getNotificationViewModelList());
        binding.recyclerView.setAdapter(adapter);
        binding.recyclerView.setLayoutManager(new LinearLayoutManager(this));
        binding.recyclerView.addItemDecoration(new DividerItemDecoration(this, LinearLayoutManager.VERTICAL));
    }

    @Override
    protected void onDestroy() {
        viewModel.destroy();
        binding.unbind();
        super.onDestroy();
    }

    private static class Adapter extends ObservableListRecyclerAdapter<NotificationViewModel, BindingHolder<ViewNotificationCellBinding>> {

        public Adapter(@NonNull ObservableList<NotificationViewModel> list) {
            super(list);
        }

        @Override
        public BindingHolder<ViewNotificationCellBinding> onCreateViewHolder(ViewGroup parent, int viewType) {
            return new BindingHolder<>(parent.getContext(), parent, R.layout.view_notification_cell);
        }

        @Override
        public void onBindViewHolder(BindingHolder<ViewNotificationCellBinding> holder, int position) {
            NotificationViewModel viewModel = getItem(position);
            ViewNotificationCellBinding itemBinding = holder.binding;
            itemBinding.setViewModel(viewModel);
            itemBinding.executePendingBindings();
        }
    }
}

File: app/src/main/java/com/u1fukui/bbs/ui/notification/NotificationListActivity.kt
package com.u1fukui.bbs.ui.notification


import android.os.Bundle
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.databinding.ObservableList
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import com.u1fukui.bbs.R
import com.u1fukui.bbs.customview.BindingHolder
import com.u1fukui.bbs.customview.ObservableListRecyclerAdapter
import com.u1fukui.bbs.databinding.ActivityNotificationListBinding
import com.u1fukui.bbs.databinding.ViewNotificationCellBinding
import com.u1fukui.bbs.repository.NotificationListRepository
import com.u1fukui.bbs.ui.BaseActivity

class NotificationListActivity : BaseActivity() {

    private val binding by lazy {
        DataBindingUtil.setContentView<ActivityNotificationListBinding>(this, R.layout.activity_notification_list)
    }

    private val viewModel: NotificationListViewModel by lazy {
        ViewModelProviders
                .of(this, NotificationListViewModel.Factory(
                        NotificationListRepository()
                ))
                .get(NotificationListViewModel::class.java)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding.viewModel = viewModel

        initToolbar(binding.toolbar, true)
        initViews()

        viewModel.start()
    }

    private fun initViews() {
        binding.recyclerView.apply {
            adapter = Adapter(viewModel.notificationBindingModelList)
            layoutManager = LinearLayoutManager(context)
            addItemDecoration(
                DividerItemDecoration(
                    context,
                    LinearLayoutManager.VERTICAL
                )
            )
        }
    }

    override fun onDestroy() {
        binding.unbind()
        super.onDestroy()
    }

    private class Adapter(list: ObservableList<NotificationBindingModel>) : ObservableListRecyclerAdapter<NotificationBindingModel, BindingHolder<ViewNotificationCellBinding>>(list) {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = BindingHolder<ViewNotificationCellBinding>(parent.context, parent, R.layout.view_notification_cell)

        override fun onBindViewHolder(holder: BindingHolder<ViewNotificationCellBinding>, position: Int) {
            holder.binding?.apply {
                bindingModel = getItem(position)
                executePendingBindings()
            }
        }
    }
}
